src/coreclr/jit/helperexpansion.cpp

```diff
@@ -442,7 +442,9 @@ PhaseStatus Compiler::fgExpandThreadLocalAccess()
         return result;
     }
 
-    if (opts.OptimizationDisabled())
+    // Always expand for NativeAOT
+    const bool isNativeAOT = IsTargetAbi(CORINFO_NATIVEAOT_ABI);
+    if (!isNativeAOT && opts.OptimizationDisabled())
```

Comment:
Why do we want this? Shouldn't we bail on MinOpts or debuggable code even for NativeAOT?

Reply:
Note to myself: Write a comment and test with MinOpts as well.

Reply:
i don't think we need to expand this for MinOpts. cc: @MichalStrehovsky, just to make sure it is ok, given there was a similar situation in https://github.com/dotnet/runtime/issues/68278#issuecomment-1543322819 where decided to always expand.

Reply:
MinOpts output:
```asm
G_M35921_IG01:  ;; offset=0x0000
       push     rbp
       push     rdi
       sub      rsp, 40
       lea      rbp, [rsp+0x30]
                                        ;; size=11 bbWeight=1 PerfScore 2.75
G_M35921_IG02:  ;; offset=0x000B
       nop
       call     CORINFO_HELP_READYTORUN_THREADSTATIC_BASE
       mov      ecx, dword ptr [rax+0x0C]
       call     Program:Consume(int)
       nop
       nop
                                        ;; size=16 bbWeight=1 PerfScore 4.75
G_M35921_IG03:  ;; offset=0x001B
       add      rsp, 40
       pop      rdi
       pop      rbp
       ret
```

Reply:
If you do not expend it in some cases, we would need to bring back the hand-emitted assembly helper (https://github.com/dotnet/runtime/pull/89472/files#diff-16ff12d6fd892f572024a7c2908c502042357d0bd176dff0036b15d7aa1f3254L240). Is that right?

Reply:
That's what my impression was, but seems it is needed only for `UseInlinedThreadStatics()` 

https://github.com/dotnet/runtime/blob/31fa0794d63a41a1b5db50bc4a5e3eadb6a786eb/src/coreclr/tools/aot/ILCompiler/Program.cs#L523-L529

which is only available when `useScanner == true` which happens in retail builds only.

https://github.com/dotnet/runtime/blob/31fa0794d63a41a1b5db50bc4a5e3eadb6a786eb/src/coreclr/tools/aot/ILCompiler/Program.cs#L303-L308

Reply:
@VSadov - is that correct?

Reply:
Threadstatics with inlined storage (no indirection to individual variables) are only possible when we have a global pre-pass, aka scanner. The scanner pass is only enabled in release, but i think it can be forced via cmd line switches. 
If such combination is real (debug + scanner), you may need to either expand or keep the helper. 
We can also agree to never produce inlined TLS form if not optimizing, regardless of scanner. Inlined form is an optimization.

Reply:
RyuJIT should ideally key this off from JitInterface - on JitInterface side we know whether the threadstatics can be inlined or not. One can also run the compiler in a mode where we optimize in RyuJIT but do not run whole program analysis so we cannot do the inline statics.

I don't view avoiding this helper as an optimization only thing. Accessing fields directly inside codegen is the norm in AOT compiled languages, irrespective of whether we optimize or not. Only .NET is so weird that sometimes we need a helper. In this case we need helper if we didn't do whole program analysis to lay out all threadstatics in the program into a linear list.

Reply:
> In this case we need helper if we didn't do whole program analysis to lay out all threadstatics in the program into a linear list.

It is a bit the other way around. 
The threadstatics that are not packed as a single linear blob almost always exist - for example threadstatics that materialize as a result of generic instantiations when T is a class. Those are always accessed via a helper. And that helper is relatively complex to be inlined by JIT. 

When we have a complete program view, some threadstatics (in reality most of them, but not all) can be optimized into single contiguous blobs with very simple access pattern. There is a much simpler helper to access these. It also can be inlined by the JIT.

> One can also run the compiler in a mode where we optimize in RyuJIT but do not run whole program analysis so we cannot do the inline statics.

In this mode there will be no threadstatics that JIT can optimize. Everything will be on the slow plan, but it would "just work".

Reply:
Basically the hand-written assembly needs to exist only if there is a mode where ILC optimizes threadstatics, but JIT does not want to.

We have 3 options:
1. JIT always optimizes inline statics (even in debug), then we do not need to keep the asm helper.
2. ILC does only produces optimized threadstatics in release and JIT always optimizes them, - no need for the asm helper. 
3. ILC produces optimized threadstatics whenever it chooses to do so, JIT optimizes them only in release, otherwise the helper is emitted/called.

I think either of these options is acceptable. We just need to decide on which one we want.

I think my _mild_ preferences would be for the '#3' as the most decoupled option that can cover any kind of combination if there is a need for that.

Reply:
The only time we will need asm helper is if is optimized by ilc (using `--scan` option) but optimization is disabled by JIT as seen in below table. 

Optimized by ILC | Optimized by JIT | Needs asm helper |
----|-----|----|
No | Yes | No
No | Yes | No
Yes | No | Yes
Yes | Yes | No

After my change in https://github.com/dotnet/runtime/pull/89472/commits/ceda4782b7beb70fb53f6bfc9f5abb448b0cb2a9, we fail if we do this:

```cmd
D:\git\runtime>ilc @"D:\git\runtime\artifacts\bin\repro\x64\Debug\compile-with-Release-libs.rsp" --scan
Unhandled exception: System.NotImplementedException: The method or operation is not implemented.
   at ILCompiler.DependencyAnalysis.ReadyToRunHelperNode.EmitInlineTLSAccess(NodeFactory factory, X64Emitter& encoder) in D:\git\runtime\src\coreclr\tools\aot\ILCompiler.Compiler\Compiler\DependencyAnalysis\Target_X64\X64ReadyToRunHelperNode.cs:line 279
   at ILCompiler.DependencyAnalysis.ReadyToRunHelperNode.EmitCode(NodeFactory factory, X64Emitter& encoder, Boolean relocsOnly) in D:\git\runtime\src\coreclr\tools\aot\ILCompiler.Compiler\Compiler\DependencyAnalysis\Target_X64\X64ReadyToRunHelperNode.cs:line 79
   at ILCompiler.DependencyAnalysis.AssemblyStubNode.GetData(NodeFactory factory, Boolean relocsOnly) in D:\git\runtime\src\coreclr\tools\Common\Compiler\DependencyAnalysis\AssemblyStubNode.cs:line 45
   at ILCompiler.DependencyAnalysis.ObjectNode.GetStaticDependencies(NodeFactory factory) in D:\git\runtime\src\coreclr\tools\Common\Compiler\DependencyAnalysis\ObjectNode.cs:line 59
```

So I think I prefer going with option #1. That's what we do for static inits too.
https://github.com/dotnet/runtime/blob/e458d68a3879d968faa1da9de38876e23935c166/src/coreclr/jit/helperexpansion.cpp#L941-L943

Reply:
`#1` seems ok too.

Reply:
> So I think I prefer going with option 1. That's what we do for static inits too.

My preference is for that one too

Reply:
Turns out that there is one scenario where we might want to retain the existing assembly i.e. "optimize-for-size". https://github.com/dotnet/runtime/issues/97168 . I don't think it is a good idea to ignore "optimize-for-size" and inline TLS access. When I tried doing that on `repro` project, it increases the object file size by 8KB. So probably I will add the assembly back.

Reply:
> it increases the object file size by 8KB

Is it object size (.obj) or final .exe size? We care about final .exe size.

Reply:
exe size increase would be somewhat unexpected. When JIT would not emit the TLS access, roughly the same code would be emitted in asm stubs. 

In theory we could emit just one asm stub and reuse it, so some size optimization is possible, but I believe we do not do that and emit a stub per call site. (historically stubs differed a bit between call sites, but not any more. For inlined TLS it is exactly the same code)

Reply:
Small correction - In cases when we need to run .cctor, the stubs are different, since they know which type, but those that do not run constructors have the same code.

Reply:
Here are the stats. For completion, I have also included the stats for `-O` to show the impact of inlining TLS

| Timeline | Compilation mode | Size of reproNative.exe (in KB) |
|---|---|---|
| Before this PR | --Os | 3981 |
| This PR | --OS | 3981 |
| This PR + ignore preferSize | --OS | 3985 |
| Before this PR | -O | 4041 |
| This PR | -O | 4044 |

Reply:
Are these numbers for release build? Release build of `Console.WriteLine("Hello world")` should be around 1MB.

Reply:
Yes, I had wrong configuration:

| Timeline | Compilation mode | Size of reproNative.exe (in KB) |
|:---:|:---:|:---:|
| Before this PR | --Os | 1109 |
| This PR | --OS | 1109 |
| This PR + ignore preferSize | --OS | 1109 |
| Before this PR | -O | 1132 |
| This PR | -O | 1132 |